[WIP] Clickhouse JDBC connector

spaghettifunk opened the pull request:
This work is based on PostgreSQL connector. It aims to solve #12440 and #9620. I am not an expert of PrestoDB and I would love some help.

Currently there is a Maven error that I can't solve and I'd like some help in testing the connector better.

highker commented on presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseClient.java:
```diff
@@ -0,0 +1,112 @@
+public class ClickhouseClient
+        extends BaseJdbcClient
+{
+    private static final String DUPLICATE_TABLE_SQLSTATE = "42P07";
```
This is the Postgres error code (https://www.postgresql.org/docs/8.2/errcodes-appendix.html). This is not necessary. If we really wanna refine errors, we can refer to Clickhouse error code or we can just throw general jdbc errors.

highker commented on presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseClient.java:
```diff
@@ -0,0 +1,112 @@
+    protected String toSqlType(Type type)
+    {
+        if (VARBINARY.equals(type)) {
+            return "bytea";
```
This also Postgres data type. Let's refer to this doc (https://clickhouse.yandex/docs/en/data_types/) for Clickhouse.

highker commented on presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseClient.java:
```diff
@@ -0,0 +1,112 @@
+    @Inject
+    public ClickhouseClient(JdbcConnectorId connectorId, BaseJdbcConfig config)
+    {
+        super(connectorId, config, "\"", new DriverConnectionFactory(new ClickHouseDriver(), config));
```
`"` is a valid identifier quote according to https://clickhouse.yandex/docs/en/query_language/syntax/. Also ` is a valid one also. Maybe we should put this as a static variable and make a comment. Picking either one is fine.

highker commented on presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseClient.java:
```diff
@@ -0,0 +1,112 @@
+            createTable(tableMetadata, tableMetadata.getTable().getTableName());
+        }
+        catch (SQLException e) {
+            if (DUPLICATE_TABLE_SQLSTATE.equals(e.getSQLState())) {
```
Remove; check the comment above.

highker commented on presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseClient.java:
```diff
@@ -0,0 +1,112 @@
+        // Clickhouse does not allow qualifying the target of a rename
+        try (Connection connection = connectionFactory.openConnection()) {
+            String sql = format(
+                    "ALTER TABLE %s RENAME TO %s",
```
Clickhouse has a different rename syntax (https://clickhouse.yandex/docs/en/query_language/misc/#rename).

zeeshanabid94 commented:
@spaghettifunk It looks like your plugin is looking good. Ping me if you need help with it.

zeeshanabid94 commented:
@spaghettifunk Just a heads up. The syntax of most sql databases is `<Database>.<Schema>.<Table>`. But click house does not have a concept of Schema. Hence, you will have to probably inject a custom query builder. The query builder is the same as the one used in `presto-jdbc`, with only the if condition on this line removed.
https://github.com/prestodb/presto/blob/master/presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/QueryBuilder.java#L125

zeeshanabid94 commented:
@spaghettifunk If you need help figuring out dependencies in `pom.xml` and `querybuilder`, you can checkout my repo.
https://github.com/zeeshanabid94/presto/tree/presto-clickhouse/presto-clickhouse/src/main/java/com.facebook.presto.plugin.clickhouse

spaghettifunk commented:
@zeeshanabid94 I added in this PR the `QueryBuilder` you built. I also fixed the comments from @highker but I feel this is still quite far to be ready :)

highker commented:
@spaghettifunk, thanks for the quick addressing on the comments. Sure, I will let you work on the PR. Ping me once it is ready.

zeeshanabid94 commented:
@spaghettifunk Good work man! Let me know if you need help.

spaghettifunk commented:
@zeeshanabid94 have you ever made a docker image of presto? I want to test the driver and i'd like to use docker for simplicity. 
Another question: how do you think I should create tests for this driver? I checked MySQL and PostgreSQL and they both have a `testing-server` that can be used to run **SmokeTests**. Any ideas how to achieve something similar?

zeeshanabid94 commented:
@spaghettifunk I have not. But it should not be hard to do that, given that you are able to make a docker file that sets up the relevant environment for presto. 
Other plugins don't use docker. Do they?

Also, I figured out how to use the query builder. `Injecting` is not the right way. Please look at my updated plugin to see how the query builder is used.
You have to add this method to your `ClickhouseClient class`
```
@Override
    public PreparedStatement buildSql(Connection connection, JdbcSplit split, List<JdbcColumnHandle> columnHandles)
            throws SQLException
    {
        return new ClickhouseQueryBuilder(identifierQuote).buildSql(
                this,
                connection,
                split.getCatalogName(),
                split.getSchemaName(),
                split.getTableName(),
                columnHandles,
                split.getTupleDomain(),
                split.getAdditionalPredicate());
    }
```

929359291 commented:
PLEASE, please help me, i have a question！clickhouse-connector read decimal column error，my clickhouse table column is decimal(10, 2)

com.facebook.presto.spi.PrestoException: DECIMAL precision must be in range [1, 38]
